### PR TITLE
Safe triangle for submenus

### PR DESCRIPTION
**Pull request by reidbarber**
Closes https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=30006503

We would like for users to be able to mouse from an item to the submenu that is opened when it was hovered. Sometimes the users mouse path might go outside of the item, which would close the submenu.

We can improve this experience by using a safe area (triangle) that users can mouse over without closing the submenu.

SafeTriangle

TODO:
- remove red background

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-spectrum/menu/src/SafeTriangle.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  }, []);
+
+  let x = 0;
+  let y = 0;
+  let width = 0;
+  let height = 0;
+  let direction: 'left' | 'right';
+  let subMenuNode = subMenuRef.current?.UNSAFE_getDOMNode() as HTMLElement;
+  if (subMenuNode) {
+    let rect = subMenuNode.getBoundingClientRect();
```
Comment (snowystinger) [inline packages/@react-spectrum/menu/src/SafeTriangle.tsx:50]: I see a lot of reading the dom in this render, this should all be moved inside an effect and the result should be stored in state

I think doing that may help with the flicker that happens some times where the triangle is the wrong size

**Review by snowystinger — COMMENTED** [review]
This is so cool!

File: packages/@react-spectrum/menu/src/SafeTriangle.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  subMenuRef: MutableRefObject<any>,
+  /** Ref for the submenu's trigger element. */
+  triggerRef: MutableRefObject<HTMLLIElement>
+}
+
+/**
+ * Allows the user to move their pointer to the sub-menu without it closing
+ * due to the mouse leaving the trigger element.
+ */
+export function SafeTriangle(props: SafeTriangleProps): ReactElement {
```
Comment (reidbarber) [inline packages/@react-spectrum/menu/src/SafeTriangle.tsx:26]: Question for the team: Would we want a react-aria hook for this? Or a RAC feature?

Comment (snowystinger) [reply packages/@react-spectrum/menu/src/SafeTriangle.tsx:26]: Good question, I'm in favor of it. Hook if possible, but we don't support the sub-menu/unavailable items in RAC yet, so might be difficult to do in this PR?

Comment (reidbarber) [reply packages/@react-spectrum/menu/src/SafeTriangle.tsx:26]: Yeah, I think we'll just include it in React Spectrum for now for that reason, but it should be easy to break out in the future for RAC.

File: packages/@react-spectrum/menu/src/SafeTriangle.tsx [hunk 3]
```diff
@@ -20,66 +20,54 @@ interface SafeTriangleProps {
...
+          top: triggerRef.current?.getBoundingClientRect().top - triggerRef.current?.parentElement.getBoundingClientRect().top,
+          left: direction === 'right' ? mouseX - left + offset : undefined,
+          height,
+          width: direction === 'right' ? left - mouseX : mouseX - right,
+          clipPath: direction === 'right' ? `polygon(100% 0%, 0% ${(100 * (mouseY - top)) / height}%, 100% 100%)` : `polygon(0% 0%, 100% ${(100 * (mouseY - top)) / height}%, 0% 100%)`
+        });
+      }
+    };
+  
+    window.addEventListener('mousemove', onMouseMove);
```
Comment (snowystinger) [inline packages/@react-spectrum/menu/src/SafeTriangle.tsx:71]: pointermove is preferred, if it should only be for mouse, you can check the event for the pointertype
not sure we still need to fall back to mousemove, pointer events are supported on all the browsers we target now

File: packages/@react-spectrum/menu/src/SafeTriangle.tsx [hunk 4]
```diff
@@ -20,66 +20,54 @@ interface SafeTriangleProps {
...
   let {
     subMenuRef,
     triggerRef
   } = props;
-  let [mousePosition, setMousePosition] = useState<{mouseX: number, mouseY: number}>({mouseX: null, mouseY: null});
-  let {mouseX, mouseY} = mousePosition;
-  let updateMousePosition = (e: MouseEvent) => {
-    setMousePosition({mouseX: e.clientX, mouseY: e.clientY});
-  };
+  let [style, setStyle] = useState<React.CSSProperties>({height: 0, width: 0});
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-spectrum/menu/src/SafeTriangle.tsx:33]: should probably set display: none on the element as well when it's not active

**Review by snowystinger — CHANGES_REQUESTED** [review]

**Comment by devongovett** [comment]
How are we going to handle this in hooks/rac implementations? Is there a way to implement it that doesn't require an extra DOM node to be added?

**Comment by reidbarber** [comment]
> How are we going to handle this in hooks/rac implementations? Is there a way to implement it that doesn't require an extra DOM node to be added?

@devongovett I was also looking at a way to do this, and thought maybe we could try to stop propagation of pointer events on the underlying menu if it the mouse is within that area, but haven't tried that yet.

**Comment by reidbarber** [comment]
@devongovett still WIP, but I think we could do something like this: [WIP: Hook for safely moving mouse for submenus](https://github.com/adobe/react-spectrum/pull/4975#top)

**Comment by reidbarber** [comment]
Closing in favor of https://github.com/adobe/react-spectrum/pull/5082

**Pull request closed without merging**